include/cetl/pf17/optional.hpp

```diff
@@ -0,0 +1,883 @@
+            this->m_engaged = true;
+        }
+        return *this;
+    }
+
+    /// Assignment 5
+    template <typename U,
+              std::enable_if_t<!detail::opt::convertible<T, optional<U>>, int> = 0,
+              std::enable_if_t<!detail::opt::assignable<T&, optional<U>>, int> = 0,
+              std::enable_if_t<std::is_constructible<T, const U&>::value, int> = 0,
+              std::enable_if_t<std::is_assignable<T&, const U&>::value, int>   = 0>
+    optional& operator=(const optional<U>& other)
+    {
+        if (this->m_engaged && other.m_engaged)
+        {
+            this->m_value = other.m_value;
+        }
+        else if (other.m_engaged)
+        {
+            new (std::addressof(this->m_value)) T(other.m_value);
+            this->m_engaged = true;
+        }
+        else
+        {
+            this->reset();
+        }
+        return *this;
+    }
+
+    /// Assignment 6
+    template <typename U,
+              std::enable_if_t<!detail::opt::convertible<T, optional<U>>, int> = 0,
+              std::enable_if_t<!detail::opt::assignable<T&, optional<U>>, int> = 0,
+              std::enable_if_t<std::is_constructible<T, U>::value, int>        = 0,
+              std::enable_if_t<std::is_assignable<T&, U>::value, int>          = 0>
+    optional& operator=(optional<U>&& other)
+    {
+        if (this->m_engaged && other.m_engaged)
+        {
+            this->m_value = std::move(other.m_value);
+        }
+        else if (other.m_engaged)
+        {
+            new (std::addressof(this->m_value)) T(std::move(other.m_value));
+            this->m_engaged = true;
+        }
+        else
+        {
+            this->reset();
+        }
+        return *this;
+    }
+
+    /// True if the optional is engaged.
+    CETL_NODISCARD constexpr bool has_value() const noexcept
+    {
+        return this->m_engaged;
+    }
+
+    /// Construct the value in-place. If the optional is already engaged, the
+    /// value is destroyed and replaced with the new value.
+    template <typename... Args>
+    T& emplace(Args&&... args) noexcept(std::is_nothrow_constructible<T, Args...>::value)
+    {
+        reset();
+        new (std::addressof(this->m_value)) T(std::forward<Args>(args)...);
+        this->m_engaged = true;
+        return this->m_value;
+    }
+    template <typename U, typename... Args>
+    T& emplace(std::initializer_list<U> il, Args&&... args)  //
+        noexcept(std::is_nothrow_constructible<T, std::initializer_list<U>, Args...>::value)
+    {
+        reset();
+        new (std::addressof(this->m_value)) T(il, std::forward<Args>(args)...);
+        this->m_engaged = true;
+        return this->m_value;
+    }
+
+    /// Swaps two optionals. If either is not engaged, acts like move assignment.
+    void swap(optional& other) noexcept(std::is_nothrow_move_constructible<T>::value&&  //
+                                            std::is_nothrow_swappable<T>::value)
```

Comment:
Yeah, this is because we had C++ extensions enabled. Both GCC and Clang expose `std::is_nothrow_swappable` in `-std=gnu++14` mode but not in the strict `-std=c++14` mode; see <https://godbolt.org/z/h9PjeE7sG>. Our CMake recipe had a statement `set(CXX_EXTENSIONS OFF)` but the correct variable for this is `CMAKE_CXX_EXTENSIONS` unless you set it as a per-target property.